hips/hip-9999.md

```diff
@@ -0,0 +1,167 @@
+---
+hip: 9999
+title: "Add descriptions to custom completions"
+authors: [ "Marc Khouzam <marc.khouzam@montreal.ca>" ]
+created: "2020-11-14"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+Provide (optional) descriptions for helm's custom completions.
+
+The goal is to help users be more efficient by providing useful information
+when doing auto-completion.
+
+## Motivation
+
+Helm currently supports auto-completion for the following shells: `bash`, `zsh`,
+`fish`. Both the `zsh` and `fish` shells support descriptions for completions.
+
+For example, in the case of `zsh`:
+```
+$ helm s<TAB>
+search  -- search for a keyword in charts
+show    -- show information of a chart
+status  -- display the status of the named release
+```
+
+Cobra automatically provides completion descriptions when completing commands
+and flags, based on the specified usage information.  However, custom
+completions that are provided by helm currently don't include any descriptions.
+For example, when completing `helm status` it is helm that provides the list of
+releases, but this list does not include descriptions.
+```
+$ helm status n<TAB>
+nginx          nginx-ingress
+```
+
+Augmenting helm's custom completions with descriptions allows providing
+additional information to help users make their selection from the list of
+completions.
+
+## Rationale
+
+The completions for some helm arguments do not always provide sufficient
+information for a user to know which value to choose.  For instance, the
+name of a release is not always informative enough for the user to know
+which one to select.  For example:
+```
+$ helm upgrade ngin<TAB>
+nginx         nginx2
+```
+Clearly, in this case, the user needs to know the details of each release
+in advance as the difference between completion choices is not very helpful.
+
+## Specification
+
+Here are proposed descriptions for the different types of custom completions:
+* chart names (for 'install', 'pull', 'show *', 'template', 'upgrade')
+    * None?
+
+    The obvious choice would be to use the description of the chart provided
+    in the index file of the repository.  However, in practice, this description
+    is often much too long to be usable for completions.
+
+    Is there other *useful* information to show? If not, it may be more useful
+    not to add a description in this case to allow to fit more chart completion
+    choices on the screen.
```

Comment:
I think I can agree with this point. The description field would be a great first choice. However, I believe there is no "hard" limit on the number of accepted characters in a description field.

What if the description was short-wrapped with ellipsis similar to what we do with `helm search repo`?

```
><> helm search repo wordpress
NAME                    CHART VERSION   APP VERSION     DESCRIPTION
stable/wordpress        9.0.3           5.3.2           DEPRECATED Web publishing platform for building...
```

Reply:
Actually, the shell does the ellipsis automatically, so I'll give it a try in the reference implementation and we can discuss how it feels during that PR review.  I'll update the HIP to propose using the description for now.

Reply:
To clarify, fish shell does an ellipsis:
```
bitnami/cassandra  (Apache Cassandra is a free and open-source distributed database manage…)
bitnami/common  (A Library Helm Chart for grouping common logic between bitnami charts. Th…)
bitnami/consul  (Highly available and distributed service discovery and key-value store de…)
bitnami/contour                                  (Contour Ingress controller for Kubernetes)
```
while `zsh` simply truncates:
```
bitnami/cassandra  -- Apache Cassandra is a free and open-source distributed database ma
bitnami/common     -- A Library Helm Chart for grouping common logic between bitnami cha
bitnami/consul     -- Highly available and distributed service discovery and key-value s
bitnami/contour    -- Contour Ingress controller for Kubernetes
```

For some chart descriptions, the amount of space available on a single line is not sufficient to convey complete information, as can be seen in the example above, especially for `consul`.